### PR TITLE
Logfile compression

### PR DESCRIPTION
**Pull request by cebe**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #13854

best to be reviewed in two steps: 7b89162 + 3fa3ba6

### CODE REVIEW

**Comment by klimov-paul** [comment]
Why `*.gz` only?
if you wish to support archiving of the files there should be option to use any archive format `*.tar`, `*.zip`, `*.rar` etc.
Hardcoding single `*.gz` can not satisfy all needs.

**Comment by klimov-paul** [comment]
Would't it be better to simply introduce an event or callback, which should be triggered after particular file has been rotated, so developer may choose, what to do with it?
For example, besides compression, some people tranfer log file to separated file storage server.

**Comment by cebe** [comment]
> Why *.gz only?

because it is directly supported by PHP, other methods would need more complex implementation or external tools.



> Hardcoding single *.gz can not satisfy all needs.

this is a start for supporting compression, `.gz` is the most common one for logs. it can be extended later.



> Would't it be better to simply introduce an event or callback, which should be triggered after particular file has been rotated, so developer may choose, what to do with it?

you can override `compressFile()` method if you want to customize behavior



> For example, besides compression, some people tranfer log file to separated file storage server.

There are tons of different ways to do this. and dependent on the case a spearate log target should be created for that.

**Comment by klimov-paul** [comment]
> There are tons of different ways to do this. and dependent on the case a spearate log target should be created for that.

I can say just the same thing for compression feature.

**Comment by klimov-paul** [comment]
> this is a start for supporting compression, .gz is the most common one for logs. it can be extended later.

That is the point: it is a start, but where it will lead? Starting support of compression log files we can end in countless request for its expansion for many purposes.
Framework either should provide sophisticated level for file archiving or do not provide it at all.

**Comment by dynasource** [comment]
based on the fact that the framework already covers log rotation, it makes sense to include compression.  It also makes sense to provide such standard GZ compression as a default.

**Comment by rob006** [comment]
We could extract compressor to separate component - basic implementation (gzip) will be provided by framework, other options may be provided by extensions.

File: framework/log/FileTarget.php [hunk 1]
```diff
@@ -137,25 +142,70 @@ protected function rotateFiles()
...
+            } else {
+                $this->rotateFile($rotateFile, $rotateTarget);
+            }
+        }
+    }
+
+    /**
+     * Rotate a single file
+     * @param string $file the file to rotate.
+     * @param string $target the target file name.
```
Comment (samdark) [inline framework/log/FileTarget.php]: Need `@since`.

File: framework/log/FileTarget.php [hunk 2]
```diff
@@ -137,25 +142,70 @@ protected function rotateFiles()
...
+     * @since 2.0.13
+     */
+    protected function compressFile($filename)
+    {
+        if (substr($filename, -3, 3) === '.gz') {
+            return $filename;
+        }
+        $gzFilename = $filename . '.gz';
+        $f = fopen($filename, 'rb');
+        $gz = gzopen($gzFilename, 'wb9');
```
Comment (samdark) [inline framework/log/FileTarget.php]: How about making compression rate and filtration flags configurable?

File: framework/log/FileTarget.php [hunk 3]
```diff
@@ -137,25 +142,70 @@ protected function rotateFiles()
...
+        while(!feof($f)) {
```
Comment (samdark) [inline framework/log/FileTarget.php]: Formatting...

File: framework/log/FileTarget.php [hunk 4]
```diff
@@ -137,25 +142,70 @@ protected function rotateFiles()
...
+            gzwrite($gz, fread($f, 8192));
```
Comment (samdark) [inline framework/log/FileTarget.php]: What is 8192?

File: framework/log/FileTarget.php [hunk 5]
```diff
@@ -137,25 +142,70 @@ protected function rotateFiles()
...
+    }
+
+
+    /**
+     * Compresses a file with gzip and renames it by appending `.gz` to the filename
+     * @param string $filename the file to compress.
+     * @return string the name of the compressed file.
+     * @since 2.0.13
+     */
+    protected function compressFile($filename)
```
Comment (samdark) [inline framework/log/FileTarget.php]: Is it possible that multiple processes call `compressFile` for the same file concurrently?

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/log/FileTarget.php (hunk 4 above)
Comment (cebe) [inline framework/log/FileTarget.php]: default chunk size from the fread docs.

File: framework/log/FileTarget.php [hunk 6]
```diff
@@ -137,25 +142,70 @@ protected function rotateFiles()
         for ($i = $this->maxLogFiles; $i >= 0; --$i) {
             // $i == 0 is the original log file
             $rotateFile = $file . ($i === 0 ? '' : '.' . $i);
-            if (is_file($rotateFile)) {
-                // suppress errors because it's possible multiple processes enter into this section
-                if ($i === $this->maxLogFiles) {
-                    @unlink($rotateFile);
+            $rotateTarget = $file . '.' . ($i + 1);
+            if (!is_file($rotateFile)) {
```
Comment (cebe) [inline framework/log/FileTarget.php]: there is no else on this one :) only if the file does not exist, something needs to be done.

File: framework/log/FileTarget.php (hunk 5 above)
Comment (cebe) [inline framework/log/FileTarget.php]: yes, same as `rotateFile()`. As far as I see they would do the compression twice. needs to be checked.

File: framework/log/FileTarget.php (hunk 6 above)
Comment (samdark) [inline framework/log/FileTarget.php]: OK. I've missed extra condition inside looking at `continue` statement. Everything's fine.

**Comment by samdark** [comment]
Does not seem we can fix it in time to get into 2.0.

**Pull request closed without merging**